src/libstore/build/local-derivation-goal.cc

```diff
@@ -803,9 +803,21 @@ void LocalDerivationGoal::startBuilder()
         userNamespaceSync.create();
 
         Path maxUserNamespaces = "/proc/sys/user/max_user_namespaces";
-        static bool userNamespacesEnabled =
-            pathExists(maxUserNamespaces)
-            && trim(readFile(maxUserNamespaces)) != "0";
+        Path procSelfNsUser = "/proc/self/ns/user";
+        static bool userNamespacesEnabled = false;
+        static bool userNamespacesTested = false;
+
+        /* Only check once */
+        if (!userNamespacesTested) {
```

Comment:
Could this be replaced by a `static` variable?

Reply:
Not sure what you mean, it is already declared as static variable in line 808, so the check should only run once (as far as I know, as I'm assuming the same behavior as in C). Do you mean a static member of `LocalDerivationGoal`?

Reply:
Yes sorry that wasn't very clear :upside_down_face: 

What I meant is moving all that logic in the initialization of the `static` variable. Something like :

```c++
static bool userNamespacesEnabled = [&](){
    if (pathExists(maxUserNamespaces)) {
        return trim(readFile(maxUserNamespaces)) != "0";
    } else {
        /* Linux kernel versions < 4.9 lack
         * /proc/sys/user/max_user_namespaces */
        return pathExists(procSelfNsUser);
    } 
}();
```

(If you're more used to C, the `[&](){ blah }()` syntax is just creating an anonymous lambda and applying it on the fly, which is the less-wrong way of initializing variables with a complex expression in C++)

Reply:
Oh, I see. Well removing the max namespaces check as commented below results in a much simpler initialization that doesn't require the lambda :-)